### PR TITLE
Make --namespace-packages the default

### PR DESCRIPTION
**Pull request by hauntsaninja**
Unfortunately, a number of tests fail. Please help fixing them :-)
Especially the ones to do with cache inconsistencies... There's at least
one known bug here: #7276

### CODE REVIEW

**Comment by JukkaL** [comment]
Great stuff! 

Some of the test failures can be tricky to debug; I can help with them.

**Comment by hauntsaninja** [comment]
Yeah, I'd love help here, maybe you could take a look at testAddedMissingStubsPackage or testDeletePackage4?

**Comment by JukkaL** [comment]
> Yeah, I'd love help here, maybe you could take a look at testAddedMissingStubsPackage or testDeletePackage4?

Sure, I should have time to look at these later this week.

**Comment by hauntsaninja** [comment]
Spent some time looking at this today. I was able to fix the "cache data discrepancy" errors.

I also spent some time looking at mypy daemon / fine grained errors, but that turned out to be a real rabbit hole. Since `--follow-imports=normal` support is explicitly experimental, I think it should be non-blocking for this PR (ie, I can make tests pass by adding `--no-namespace-packages`).

**Comment by JukkaL** [comment]
Sorry, I lost of track of this PR. Is there anything I can do to help move this forward?

**Comment by nipunn1313** [comment]
#11257, #11259, and #11261 should all help unblock this

**Comment by ethanhs** [comment]
We should probably add https://github.com/python/mypy/issues/10310 and other exclude usability improvements before putting this in a release because it seems that this is picking up tools/test/misc directories that people may not want checked.

**Comment by Akuli** [comment]
It just feels wrong to have one global setting that affects every library you import, regardless of whether it's a dependency of a dependency, or a part of the project are working on where you forgot to put `__init__.py`.

**Comment by hauntsaninja** [comment]
Thanks to Nipunn's efforts CI is now green. There's still some stuff to be fixed, see https://github.com/python/mypy/issues/11259

File: mypy/build.py
Comment (nipunn1313) [inline mypy/build.py:1326]: this seems unrelated? Useful - but not related.

**Comment by nipunn1313** [comment]
#11322 - is the remaining daemon issue.

I don't think we should make it a blocker as it only affects the rare scenario in which you delete an `__init__.py` file to intentionally convert into a namespace package. Workaround exists (restart daemon).

Code looks good to me!

**Review by nipunn1313 — APPROVED** [review]

File: mypy/build.py
Comment (hauntsaninja) [reply mypy/build.py:1326]: I think there was a change here that was needed, but then got merged in another PR

**Comment by JukkaL** [comment]
Great, let's get this merged! I'd like to play around with this a little before merging, since this kind of a big change, but don't wait for me if I get distracted and you don't hear anything from me in a few days.

File: mypy/build.py
Comment (nipunn1313) [reply mypy/build.py:1326]: Yeah - I think remove this from this commit/PR before landing.
Better to keep the diff scoped to what's necessary.

**Comment by nipunn1313** [comment]
Hey friends! Can we give this one a rebase and try to get it in? Thanks!

(I've been blocking https://github.com/python/typeshed/pull/6106 on this)

**Comment by 97littleleaf11** [comment]
cc @JukkaL @hauntsaninja

**Comment by hauntsaninja** [comment]
Nipunn / typeshed are blocked on us and this change is ready, so we probably shouldn't delay past the next release. But given the breakage this will cause to people's setups I'd vaguely prefer it if @JukkaL were to hit merge.

I'm also curious if people have thoughts on how we could make this change go smoother. A couple thoughts:
- Maybe we can make "Duplicate module named ..." not cause fatal build errors, or make the error note give clearer advice?
- Similar thing for "Source file found twice under different module names"... I added a link to the docs in #11270, but we could more clearly say that the action to be taken is probably a) putting something on mypy path, b) adding an `__init__.py`, or c) using `exclude`
- Fix https://github.com/python/mypy/issues/11322
- Fix https://github.com/python/mypy/issues/9716
- If there are other breaking changes we could bundle with this in the next release (to avoid a drip-drip-drip of breaking changes). Candidates changes include: https://github.com/python/mypy/issues/8584#issuecomment-927361405

**Comment by srittau** [comment]
I think @JukkaL gave his approval in #12021.

File: mypy/main.py
Comment (97littleleaf11) [inline mypy/main.py:520]: Please also update the documentation. 
> https://mypy.readthedocs.io/en/stable/command_line.html#import-discovery
> https://mypy.readthedocs.io/en/stable/config_file.html?highlight=namespace-packages#confval-namespace_packages

Comment (hauntsaninja) [reply mypy/main.py:520]: Thanks, updated!

**Comment by ethanhs** [comment]
Since we are already going to potentially have breakage due to https://github.com/python/mypy/pull/11143, and I plan on working on https://github.com/python/mypy/issues/8545 in the next week, perhaps this release is a good time to push this change?

**Comment by ofek** [comment]
Can this be merged?

**Comment by nipunn1313** [comment]
Wanted to check in here to see if its possible to go forward with this.

**Comment by AlexWaygood** [comment]
Looks like this caused the selfcheck to start failing: https://github.com/python/mypy/actions/runs/3133286021/jobs/5086495727

**Comment by AlexWaygood** [comment]
> Looks like this caused the selfcheck to start failing on master: https://github.com/python/mypy/actions/runs/3133286021/jobs/5086495727

Fixed in #13748

**Pull request merged**
